common/src/Microsoft.Azure.IIoT.Hub.Module.Client/src/Default/MqttClient/MqttClientAdapter.cs

```diff
@@ -0,0 +1,616 @@
+            /// </summary>
+            /// <param name="client">MQTT server client.</param>
+            /// <param name="product">Custom product information.</param>
+            /// <param name="cs">Connection string for the MQTT server.</param>
+            /// <param name="deviceId">Id of the device.</param>
+            /// <param name="telemetryTopicTemplate">Telemetry topic template.</param>
+            /// <param name="timeout">Timeout used for operations.</param>
+            /// <param name="retry">Retry policy used for operations.</param>
+            /// <param name="onConnectionLost">Handler for when the MQTT server connection is lost.</param>
+            /// <param name="logger">Logger used for operations</param>
+            /// <returns></returns>
+            public static async Task<IClient> CreateAsync(
+                IManagedMqttClient client,
+                string product,
+                MqttClientConnectionStringBuilder cs,
+                string deviceId,
+                string telemetryTopicTemplate,
+                TimeSpan timeout,
+                IRetryPolicy retry,
+                Action onConnectionLost,
+                ILogger logger) {
+                var options = new MqttClientOptionsBuilder()
+                    .WithClientId(cs.DeviceId)
+                    .WithTcpServer(tcpOptions => {
+                        tcpOptions.Server = cs.HostName;
+                        tcpOptions.Port = cs.Port;
+
+                        if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux)) {
+                            tcpOptions.BufferSize = 64 * 1024;
+                        }
+                    });
+
+                if (!string.IsNullOrWhiteSpace(cs.Username) && !string.IsNullOrWhiteSpace(cs.Password)) {
+                    options = options.WithCredentials(cs.Username, cs.Password);
+                }
+
+                var tls_options = new MqttClientOptionsBuilderTlsParameters {
+                    UseTls = cs.UsingIoTHub,
+                    SslProtocol = SslProtocols.Tls12,
+                    IgnoreCertificateRevocationErrors = true
+                };
+
+                if (cs.UsingX509Cert) {
+                    tls_options.Certificates = new List<X509Certificate> { cs.X509Cert };
+                }
+                options = options.WithTls(tls_options);
+
+                // Use MQTT 5.0 if message expiry interval is set.
+                if (cs.MessageExpiryInterval != null) {
+                    options = options.WithProtocolVersion(MqttProtocolVersion.V500);
+                }
+
+                var adapter = new MqttClientAdapter(
+                    client,
+                    product,
+                    deviceId,
+                    timeout,
+                    retry,
+                    onConnectionLost,
+                    cs.MessageExpiryInterval,
+                    cs.UsingIoTHub ? MqttQualityOfServiceLevel.AtLeastOnce : MqttQualityOfServiceLevel.ExactlyOnce,
```

Comment:
Do we really want to use QoS2 if connected to another MQTT broker than IoT Hub?

Reply:
don't think so. but not my change. will go to QoS1